### PR TITLE
Multiple updates (multi-OS pool, service logging, exception handling,…

### PR DESCRIPTION
**Pull request by EarthmanT**
… new API endpoints)

Multiple updates (multi-OS pool, service logging, exception handling, new API endpoints)

Testing REST updates

Testing REST updates

Fixing service liveness test

Fixing service liveness test

Fixing service liveness test

Fixing accidental pushes

Fixing VIRTUALENV search

Fixing VIRTUALENV search

Fixing VIRTUALENV search

Fixed service issue

Fixed service issue

Fixed tests

Fixing tests

Reworking service, tests

Reworking service, tests

Reworking service, tests

Reworking service, tests

Fixing tests

Fixing tests

Removing requirement for config YAML file

Fixing working_directory input

creating non-daemon service option

creating non-daemon service option

creating non-daemon service option

creating non-daemon service option

creating non-daemon service option

creating non-daemon service option

Updating API request data support

Fixed tests

Fixed defaults issue

Adding update functionality

All tests are now passing

Changing how keys are added

Testing new service logger

Testing new service logger

Added Flask, Gunicorn logging

Simplifying host scanning

Added flast-restful exception handling

Adding tcp scan info

Preparing for merge

CFY-4959 - Rewrote API server, added service logging, changed backend to TinyDB, added mixed-OS pool support.

add build auto-creation on new tag

Fixing tests, updating test package versions

Adding init script, debug loggers, and new lifecycle scripts

Added logger test. Fixed exception handling

Fixing styling & exception handling

Fixing service liveness test

Fixing service liveness test

Fixing service liveness test

Fixing accidental pushes

Fixing VIRTUALENV search

Fixed service issue

Fixing tests

Reworking service, tests

Reworking service, tests

Removing requirement for config YAML file

Fixing working_directory input

creating non-daemon service option

creating non-daemon service option

Fixed defaults issue

Testing new service logger

Added flast-restful exception handling

Adding tcp scan info

Preparing for merge

Fixed broken POST /host/allocate with OS specified.  Fixed case where host is allocated but scanning fails.

add build auto-creation on new tag

Fixing tests, updating test package versions

Adding init script, debug loggers, and new lifecycle scripts

Added logger test. Fixed exception handling

Fixing styling & exception handling

Testing REST updates

Fixing service liveness test

Fixing service liveness test

Fixing service liveness test

Fixing accidental pushes

Fixing VIRTUALENV search

Fixing VIRTUALENV search

Fixing VIRTUALENV search

Fixed service issue

Fixed service issue

Fixed tests

Fixing tests

Reworking service, tests

Reworking service, tests

Reworking service, tests

Reworking service, tests

Fixing tests

Fixing tests

Removing requirement for config YAML file

Fixing working_directory input

creating non-daemon service option

creating non-daemon service option

creating non-daemon service option

creating non-daemon service option

creating non-daemon service option

creating non-daemon service option

Updating API request data support

Fixed tests

Fixed defaults issue

Adding update functionality

All tests are now passing

Changing how keys are added

Testing new service logger

Testing new service logger

Added Flask, Gunicorn logging

Simplifying host scanning

Added flast-restful exception handling

Adding tcp scan info

Preparing for merge

CFY-4959 - Rewrote API server, added service logging, changed backend to TinyDB, added mixed-OS pool support.

Fixing service liveness test

Fixing service liveness test

Fixing accidental pushes

Fixing VIRTUALENV search

Fixing VIRTUALENV search

Fixing VIRTUALENV search

Fixed service issue

Fixed tests

Fixing tests

Reworking service, tests

Reworking service, tests

Removing requirement for config YAML file

creating non-daemon service option

creating non-daemon service option

Fixed defaults issue

All tests are now passing

Testing new service logger

Testing new service logger

Added flast-restful exception handling

Adding tcp scan info

Preparing for merge

add build auto-creation on new tag

Fixing tests, updating test package versions

Adding init script, debug loggers, and new lifecycle scripts

Added logger test. Fixed exception handling

Fixing styling & exception handling

Added init script, debug loggers, new lifecycle scripts, and logger test

Fixing tests, updating test package versions

add build auto-creation on new tag

Fixing tests, updating test package versions

Adding init script, debug loggers, and new lifecycle scripts

Added logger test. Fixed exception handling

Fixing styling & exception handling

Added init script, debug loggers, new lifecycle scripts, and logger test

add build auto-creation on new tag

Fixing tests, updating test package versions

Adding init script, debug loggers, and new lifecycle scripts

Added logger test. Fixed exception handling

Fixing styling & exception handling

add build auto-creation on new tag

Fixing tests, updating test package versions

Adding init script, debug loggers, and new lifecycle scripts

Added logger test. Fixed exception handling

Fixing styling & exception handling

Added init script, debug loggers, new lifecycle scripts, and logger test

Fixing tests, updating test package versions

Adding init script, debug loggers, and new lifecycle scripts

Added logger test. Fixed exception handling

Fixing styling & exception handling

add build auto-creation on new tag

Fixing tests, updating test package versions

Testing REST updates

Fixing service liveness test

Fixing service liveness test

Fixing service liveness test

Fixing accidental pushes

Fixing VIRTUALENV search

Fixing VIRTUALENV search

Fixing VIRTUALENV search

Fixed service issue

Fixed service issue

Fixed tests

Fixing tests

Reworking service, tests

Reworking service, tests

Reworking service, tests

Reworking service, tests

Fixing tests

Fixing tests

Removing requirement for config YAML file

Fixing working_directory input

creating non-daemon service option

creating non-daemon service option

creating non-daemon service option

creating non-daemon service option

creating non-daemon service option

creating non-daemon service option

Updating API request data support

Fixed tests

Fixed defaults issue

Adding update functionality

All tests are now passing

Changing how keys are added

Testing new service logger

Testing new service logger

Added Flask, Gunicorn logging

Simplifying host scanning

Added flast-restful exception handling

Adding tcp scan info

Preparing for merge

CFY-4959 - Rewrote API server, added service logging, changed backend to TinyDB, added mixed-OS pool support.

add build auto-creation on new tag

Fixing tests, updating test package versions

Fixing service liveness test

Fixing VIRTUALENV search

Fixing VIRTUALENV search

Fixed service issue

Fixed tests

Fixing tests

Reworking service, tests

Reworking service, tests

Removing requirement for config YAML file

Fixing working_directory input

creating non-daemon service option

creating non-daemon service option

Fixed defaults issue

Testing new service logger

Testing new service logger

Added flast-restful exception handling

Adding tcp scan info

Preparing for merge

Fixed broken POST /host/allocate with OS specified.  Fixed case where host is allocated but scanning fails.

Fixing tests

Fixed missing info from problem squash

### CODE REVIEW

File: cloudify_hostpool/rest/backend.py [hunk 1]
```diff
@@ -1,164 +1,370 @@
...
-                _create_indicator()
+    @staticmethod
+    def validate_defaults(defaults):
+        '''Validates default data'''
+        if not defaults:
+            return
+        if not isinstance(defaults, dict):
+            raise exceptions.ConfigurationError(
+                'Defaults must be a valid JSON Object')
+        if defaults.get('platform'):
```
Comment (EarthmanT) [inline cloudify_hostpool/rest/backend.py]: this can never run.

File: cloudify_hostpool/rest/backend.py [hunk 2]
```diff
@@ -1,164 +1,370 @@
...
+            raise exceptions.ConfigurationError(
+                'Default "platform" is not allowed')
+        if defaults.get('endpoint') and \
```
Comment (EarthmanT) [inline cloudify_hostpool/rest/backend.py]: i don't think this can ever happen either. see line 58

File: cloudify_hostpool/rest/backend.py [hunk 3]
```diff
@@ -1,164 +1,370 @@
...
+           not isinstance(defaults.get('endpoint'), dict):
+            raise exceptions.ConfigurationError(
+                'Default "endpoint" must be a valid JSON Object')
+        if defaults.get('endpoint') and \
```
Comment (EarthmanT) [inline cloudify_hostpool/rest/backend.py]: same here.

File: cloudify_hostpool/rest/backend.py [hunk 4]
```diff
@@ -1,164 +1,370 @@
...
+# HostReconciler
+# - Handles any host entry duplication cases
 
-# if this file exists, loading of the pool will not take place.
-# note that this means that multiple instances of this application that
-# execute from the same directory will only load the pool the first
-# time, and all other instances will use the previously loaded pool.
-INDICATOR = 'host-pool-loaded-indicator'
 
+class HostAlchemist(object):
```
Comment (EarthmanT) [inline cloudify_hostpool/rest/backend.py]: does anything every inherit this?

File: cloudify_hostpool/rest/backend.py [hunk 5]
```diff
@@ -1,164 +1,370 @@
...
-                    {'reserved': False,
-                     'alive': False})
+        for host_id in self.get_unallocated_host_ids():
+            # Allocate the host
+            self.logger.debug('Trying host #{0}'.format(host_id))
+            self.storage.update_host(host_id, {'allocated': True})
+            # Refresh our data
+            host = self.storage.get_host(host_id)
+            # Enforce any user-defined requests
+            if requested_os and requested_os != host['os']:
```
Comment (EarthmanT) [inline cloudify_hostpool/rest/backend.py]: this is a long control flow here. For readability, I prefer to break into
```
if self.check_os(...) and if self.host_port_scan(...):
    return self.storage.get_host(host_id)
else:
    self.storage.update_host(host_id, {'allocated': False})
    raise exceptions.NoHostAvailableException()
```

File: examples/local-blueprint/local-blueprint.yaml [hunk 6]
```diff
@@ -1,20 +1,31 @@
-tosca_definitions_version: cloudify_dsl_1_0
+tosca_definitions_version: cloudify_dsl_1_2
 
 imports:
-  - http://www.getcloudify.org/spec/cloudify/3.2rc1/types.yaml
-  - https://raw.githubusercontent.com/cloudify-cosmo/cloudify-host-pool-service/master/host-pool-service.yaml
+  - http://www.getcloudify.org/spec/cloudify/3.3.1/types.yaml
+  - https://raw.githubusercontent.com/01000101/cloudify-host-pool-service/CFY-4959/host-pool-service.yaml
```
Comment (EarthmanT) [inline examples/local-blueprint/local-blueprint.yaml]: lets make sure to update this before merge

Comment (01000101) [inline examples/local-blueprint/local-blueprint.yaml]: We can't since it will break tests.

File: cloudify_hostpool/rest/backend.py (hunk 1 above)
Comment (01000101) [inline cloudify_hostpool/rest/backend.py]: True, if you are to assume that TestBackend::parse() is the only entry point that will ever be used.  I'd assume someone, at some point, will directly use the method and we want that check in place.

File: cloudify_hostpool/rest/backend.py (hunk 2 above)
Comment (01000101) [inline cloudify_hostpool/rest/backend.py]: Same as above.

File: cloudify_hostpool/rest/backend.py (hunk 3 above)
Comment (01000101) [inline cloudify_hostpool/rest/backend.py]: Likewise

File: cloudify_hostpool/rest/backend.py (hunk 4 above)
Comment (01000101) [inline cloudify_hostpool/rest/backend.py]: No, does it need to? I see no harm in allowing to be re-used later vs. having one giant, monolithic RestBackend class when the functionality differs.

File: cloudify_hostpool/rest/backend.py (hunk 5 above)
Comment (01000101) [inline cloudify_hostpool/rest/backend.py]: Good point, thanks.

File: cloudify_hostpool/rest/backend.py (hunk 1 above)
Comment (EarthmanT) [inline cloudify_hostpool/rest/backend.py]: fair enough

File: cloudify_hostpool/rest/backend.py (hunk 2 above)
Comment (EarthmanT) [inline cloudify_hostpool/rest/backend.py]: fair enough

File: cloudify_hostpool/rest/backend.py (hunk 3 above)
Comment (EarthmanT) [inline cloudify_hostpool/rest/backend.py]: fair enough

File: cloudify_hostpool/rest/backend.py (hunk 4 above)
Comment (EarthmanT) [inline cloudify_hostpool/rest/backend.py]: Just asking..not looking for changes

File: examples/local-blueprint/local-blueprint.yaml (hunk 6 above)
Comment (EarthmanT) [inline examples/local-blueprint/local-blueprint.yaml]: i'll take care of it after tests pass

**Pull request merged**